### PR TITLE
Prevent inner-object access of non existent attributes via [] or dig

### PR DESCRIPTION
**Pull request by tom-lord**
An experimental idea I had, after encountering this bug: https://github.com/carwow/quotes_site/pull/14658#discussion_r579090860

Previous behaviour:

```
> open_object = RestfulResource::OpenObject.new(foo: 123)
=> #<RestfulResource::OpenObject:0x000055efba1e0a78 @inner_object=#<OpenStruct foo=123>>

> open_object.foo
=> 123

> open_object.bar
NoMethodError: undefined method `bar' for #<RestfulResource::OpenObject:0x000055efba1e0a78>
from /usr/local/bundle/gems/restful_resource-2.10.3/lib/restful_resource/open_object.rb:11:in `method_missing'

> open_object[:bar]
=> nil

> open_object.dig(:bar)
=> nil
```

New behaviour:
```
> open_object = RestfulResource::OpenObject.new(foo: 123)
=> #<RestfulResource::OpenObject:0x00007fe3dd253640 @inner_object=#<RestfulResource::StrictOpenStruct foo=123>>

> open_object.foo
=> 123

> open_object.bar
=> NoMethodError (undefined method `bar' for #<RestfulResource::OpenObject:0x00007fe3dd253640>)

> open_object[:bar]
=> NoMethodError (undefined method `[]' for #<RestfulResource::OpenObject:0x00007fe3dd253640>)

> open_object.dig(:bar)
=> NoMethodError (undefined method `dig' for #<RestfulResource::OpenObject:0x00007fe3dd253640>)
```

I want to see an error raised on those last two scenarios, just like calling `open_object.bar`, to prevent accidental non-existent-attribute access from silently failing.

### CODE REVIEW

**Comment by tom-lord** [comment]
I tried running this new version in `research_site` locally and got a long list of spec failures -- but hopefully they're all fixable!!

However, if it proves too painful to fix all errors at once, we could try replacing the above with some sort of `deprecation warning` instead of a hard-failure?

File: lib/restful_resource/strict_open_struct.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module RestfulResource
+  class StrictOpenStruct < ::OpenStruct
+    undef_method :dig, :[]
```
Comment (rafaelmazza) [inline lib/restful_resource/strict_open_struct.rb:3]: I would imagine that we use `dig` and `[]` in many places so maybe doing it as a deprecation warning as you said to start with would be less annoying? And then we can create some TI cards for that.

**Review by rafaelmazza — APPROVED** [review]

**Comment by tom-lord** [comment]
We've agreed to try it this way, but maybe re-implement as a deprecation instead, if this proved to be a blocker for updating in other projects.

**Comment by CGA1123** [comment]
What were the reasons for releasing this breaking change as a minor version bump?

**Pull request merged**